### PR TITLE
Fix problems with GC compaction and non-consecutive function slots

### PR DESCRIPTION
**Pull request by Ekdohibs**
If function slots were not consecutive, the GC would previously incorrectly scan them during compaction while trying to revert infix pointers.
Note: due to an invariant in slot_offsets, this can't happen with the existing compiler.

### CODE REVIEW

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 1]
```diff
@@ -280,6 +280,8 @@ end = struct
         else if starting_offset = slot_offset
         then acc
         else
+          (* The space between slot offsets has to be padded with tagged 0, as
+             it is scanned by the GC during compaction. *)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:284]: I think we should specify here that the constraint is not only that the value put here has to be scannable, but that it has to be a very specific value, and add a link/comment to the relevant part of the compaction C code.

**Review by Gbury — APPROVED** [review]

**Comment by mshinwell** [comment]
(The precondition for this bug cannot be satisfied with the current compiler.)

**Comment by mshinwell** [comment]
Does this need a fix for runtime5?

**Comment by Ekdohibs** [comment]
I think this doesn't need a fix for runtime5, but I just noticed `extern.c` might need a fix as well; I'll look.

**Review by Gbury — APPROVED** [review]
Looks good, we should also probably make a note somewhere that compaction in runtime5 will need to be adjusted when it is merged ?

**Review by lthls — APPROVED** [review]
Approving the latest changes

**Comment by Ekdohibs** [comment]
Compaction in runtime5 has already been merged, and works differently to runtime4 so doesn't need to be changed.

**Pull request merged**